stdlib/runtime/numbers.gr

```diff
@@ -1365,94 +1390,84 @@ export let rec coerceNumberToRational = (x: Number) => {
       throw Exception.NumberNotRational
     }
   }
-  let ret = WasmI32.toGrain(result): Rational
-  if (WasmI32.fromGrain(ret) != WasmI32.fromGrain(x)) {
-    Memory.decRef(WasmI32.fromGrain(x))
-    void
-  }
-  Memory.decRef(WasmI32.fromGrain(coerceNumberToRational))
-  ret
+  WasmI32.toGrain(result): Rational
 }
 
-export let rec coerceNumberToFloat32 = (x: Number) => {
+@unsafe
+export let coerceNumberToFloat32 = (x: Number) => {
   let x = WasmI32.fromGrain(x)
   let result = if (
     !isSimpleNumber(x) &&
     WasmI32.eq(boxedNumberTag(x), Tags._GRAIN_FLOAT32_BOXED_NUM_TAG)
   ) {
-    // avoid extra malloc
+    // avoid extra malloc and prevent x from being freed
+    Memory.incRef(x)
     x
   } else {
     newFloat32(coerceNumberToWasmF32(WasmI32.toGrain(x): Number))
   }
-  let ret = WasmI32.toGrain(result): Float32
-  if (WasmI32.fromGrain(ret) != WasmI32.fromGrain(x)) {
-    Memory.decRef(WasmI32.fromGrain(x))
-    void
-  }
-  Memory.decRef(WasmI32.fromGrain(coerceNumberToFloat32))
-  ret
+  WasmI32.toGrain(result): Float32
 }
 
-export let rec coerceNumberToFloat64 = (x: Number) => {
+@unsafe
+export let coerceNumberToFloat64 = (x: Number) => {
   let x = WasmI32.fromGrain(x)
   let result = if (
     !isSimpleNumber(x) &&
     WasmI32.eq(boxedNumberTag(x), Tags._GRAIN_FLOAT64_BOXED_NUM_TAG)
   ) {
-    // avoid extra malloc
+    // avoid extra malloc and prevent x from being freed
+    Memory.incRef(x)
     x
   } else {
     newFloat64(coerceNumberToWasmF64(WasmI32.toGrain(x): Number))
   }
-  let ret = WasmI32.toGrain(result): Float64
-  if (WasmI32.fromGrain(ret) != WasmI32.fromGrain(x)) {
-    Memory.decRef(WasmI32.fromGrain(x))
-    void
-  }
-  Memory.decRef(WasmI32.fromGrain(coerceNumberToFloat64))
-  ret
+  WasmI32.toGrain(result): Float64
 }
 
-// These coerceXX functions don't need Memory.decRef calls since they simply return their argument
-export let rec coerceInt32ToNumber = (x: Int32) => {
-  let ret = WasmI32.toGrain(WasmI32.fromGrain(x)): Number
-  Memory.decRef(WasmI32.fromGrain(coerceInt32ToNumber))
-  ret
+@unsafe
+export let coerceInt32ToNumber = (x: Int32) => {
+  let x = WasmI32.fromGrain(x)
+  Memory.incRef(x)
```

Comment:
Want to add the same comments about avoiding x not being freed to these?

Reply:
Will do!